ProjFS.Mac/PrjFSKext/KauthHandler.cpp

```diff
@@ -618,7 +618,8 @@ static int HandleFileOpOperation(
             }
         }
     }
-    else if (KAUTH_FILEOP_CLOSE == action)
+    else if (KAUTH_FILEOP_CLOSE == action ||
+             KAUTH_FILEOP_OPEN == action)
     {
         currentVnode = reinterpret_cast<vnode_t>(arg0);
         putCurrentVnode = false;
```

Comment:
If we remove `if (vnode_isdir(currentVnode))` from the `KAUTH_FILEOP_OPEN` path will it allow us to receive new file notifications for directories? (it would be awesome if it did 😃 )

Reply:
Looks like it.  You see the notifications when you try to access the directory (not on mkdir).  What are you thinking here?

Reply:
> Looks like it. You see the notifications when you try to access the directory (not on mkdir). What are you thinking here?

Ahh, okay, I was wondering if we'd get it on mkdir.  Currently we detect new directories when a new file is created inside of them, and I wasn't sure if the `KAUTH_FILEOP_OPEN` notification would allow us to detect those new directories earlier.

I don't think it's something we need to change now, but we could consider changing when we trigger the notification for directories (so that they'll happen even without a file being created inside of them).

Reply:
Sounds good!, @pmj mentioned we will get a KAUTH_VNODE_ADD_SUBDIRECTORY on the parent folder.  Thats another option we could look into.

Reply:
Just to expand on that: the FILEOP open/close events will almost certainly only be delivered when the corresponding syscalls are made, and directories are only `open()`ed when *listing* directory contents. `mkdir` does not produce a file descriptor, and so does not act like an `open()` call in any way.

There is definitely a [`KAUTH_VNODE_ADD_SUBDIRECTORY` auth query](https://github.com/apple/darwin-xnu/blob/a449c6a3b8014d9406c2ddbdc81795da24aa7443/bsd/vfs/vfs_subr.c#L6399), but [no FILEOP event at all produced by `mkdir` syscalls](https://github.com/apple/darwin-xnu/blob/a449c6a3b8014d9406c2ddbdc81795da24aa7443/bsd/vfs/vfs_syscalls.c#L7997) - although it is [reported to FSEvents](https://github.com/apple/darwin-xnu/blob/a449c6a3b8014d9406c2ddbdc81795da24aa7443/bsd/vfs/vfs_syscalls.c#L8016). I don't know to what extent [FSEvents](https://developer.apple.com/library/archive/documentation/Darwin/Conceptual/FSEvents_ProgGuide/UsingtheFSEventsFramework/UsingtheFSEventsFramework.html) have been investigated for the change detection aspect of VFS4G; it obviously is no good for hydration etc. but should give fewer false positives than vnode listeners. They are of course entirely async, but should arrive in a more timely manner than `KAUTH_FILEOP_CLOSE` at any rate.